src/Extensions/CompanionServer.cpp

```diff
@@ -18,148 +18,146 @@
 #include "Extensions/CompanionServer.hpp"
 #include "Core/EventLogger.hpp"
 #include "Core/MessageLogger.hpp"
+#include "qhttpfwd.hpp"
+#include "qhttpserver.hpp"
+#include "qhttpserverconnection.hpp"
+#include "qhttpserverrequest.hpp"
+#include "qhttpserverresponse.hpp"
 #include <QJsonArray>
 #include <QJsonObject>
-#include <QTcpServer>
-#include <QTcpSocket>
+
+#define SAFEILOG(x, y)                                                                                                 \
+    if (log)                                                                                                           \
+        log->info(x, y);
+
+#define SAFEELOG(x, y)                                                                                                 \
+    if (log)                                                                                                           \
+        log->error(x, y);
+
+#define SAFEWLOG(x, y)                                                                                                 \
+    if (log)                                                                                                           \
+        log->warn(x, y);
 
 namespace Extensions
 {
-CompanionServer::CompanionServer(int port)
+CompanionServer::CompanionServer(int port, QObject *parent) : QObject(parent)
 {
-    updatePort(port);
+    startListeningOn(port);
 }
 
 void CompanionServer::setMessageLogger(MessageLogger *log)
 {
     this->log = log;
 }
 
+bool CompanionServer::startListeningOn(int port)
+{
+    if (server)
+        delete server;
+    server = new qhttp::server::QHttpServer(this);
+    server->listen(QString::number(port),
+                   [&, this](qhttp::server::QHttpRequest *req, qhttp::server::QHttpResponse *res) {
+                       LOG_INFO("\n--> " << req->methodString() << " : " << qPrintable(req->url().toString().toUtf8()));
+                       req->headers().forEach(
+                           [](auto iter) { LOG_INFO(iter.key().constData() << " : " << iter.value().constData()); });
+
+                       QString methodType = req->methodString();
+                       bool isJson = req->headers().keyHasValue("content-type", "application/json");
+                       req->collectData();
+
+                       req->onEnd([res, req, methodType, isJson, this]() {
+                           res->addHeader("connection", "close");
+                           res->addHeader("pragma", "no-cache");
+
+                           if (methodType != "POST")
+                           {
+                               SAFEWLOG("Companion", tr("Discarded %1 request").arg(methodType));
+                               res->setStatusCode(qhttp::ESTATUS_NOT_ACCEPTABLE);
+                               res->end();
+                               return;
+                           }
```

Comment:
What does this error mean? When does it happen? Why is it a warning?

Reply:
When a GET/PUT/DELETE or any other non POST request is made to the server port. We Gracefully ignore it and issue a warning. We could have ignored it all together even without logging a warning but if this happens user should know that something else in their computer is making some request to the Port which is suppose to listen for only Competitive Companion POST data.

Reply:
One way to trigger this warning is that you could visit `localhost:10045` (port of competitive companion) from your web browser. It sends a GET request and CP Editor will issue the warning. From development point of view it is another way to check if server is actually listening on the port.